nmdc_schema/migrators/migrator_from_X_to_PR31.py

```diff
@@ -0,0 +1,79 @@
+from nmdc_schema.migrators.migrator_base import MigratorBase
+from nmdc_schema.migrators.adapters.adapter_base import AdapterBase
+from difflib import SequenceMatcher
+
+class Migrator(MigratorBase):
+    r"""
+    Migrates data from X to PR31, removes used slot from WorkflowExecution subclasses and checks that the 
+    value in the used slot on the WorkflowExecution classes matches the value on the DataGeneration 
+    instances in the instrument_name slot.
+    """
+
+    _from_version = "X"
+    _to_version = "PR31"
+
+    def upgrade(self):
+        r"""Migrates the database from conforming to the original schema, to conforming to the new schema."""
+        
+        workflow_execution_collection_names = [
+            "mags_activity_set",
+            "metabolomics_analysis_activity_set",
+            "metagenome_annotation_activity_set",
+            "metagenome_assembly_set",
+            "metagenome_sequencing_activity_set",
+            "metatranscriptome_activity_set",
+            "nom_analysis_activity_set",
+            "omics_processing_set",
+            "read_based_taxonomy_analysis_activity_set",
+            "read_qc_analysis_activity_set"
+            "metaproteomics_analysis_activity_set"   
+        ]
+        
+        for collection_name in workflow_execution_collection_names:
+            self.adapter.process_each_document(
+                collection_name=collection_name,
+                pipeline=[self.remove_used_slot],
+            )
+    
+    def preprocess_string(self, s):
+
+        return s.replace(" ", "").replace("_","").replace("-","")
+
+    def remove_used_slot(self, doc: dict) -> dict:
+        r"""
+        Removes the used slot from WorkflowExecution subclasses if the value matches the 
+        instrument_name slot from the corresponding OmicsProcessing document by the longest
+        common sequence.
+
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {'omics_processing_set':[{'id':'nmdc:omcp-123', 'instrument_name':'nmdc:wfc-456'}]}  # in this example, our data store is a Python dictionary
+        >>> adapter = DictionaryAdapter(database=database)
+        >>> m = Migrator(adapter=adapter)
+        >>> m.remove_used_slot({'id': 'nmdc:metab-123', 'used': 'nmdc:wfc-456', 'was_informed_by': 'nmdc:omcp-123'})
+        {'id': 'nmdc:metab-123', 'was_informed_by': 'nmdc:omcp-123'}
+        """
+        
+        if "used" in doc:
+            omics_processing_doc = self.adapter.get_document_having_value_in_field(
+                collection_name="omics_processing_set", field_name="id", value=doc["was_informed_by"]
```

Comment:
When I ran this migrator on a copy of the production database, this line of code produced a `KeyError`. ⚠️ 

![image](https://github.com/microbiomedata/berkeley-schema-fy24/assets/134325062/f1622f9d-73ea-4913-bd5f-2bf025642209)

This line of code tries to use the `"was_informed_by"` key on the dictionary passed in. If the dictionary doesn't have that key, a `KeyError` is raised.

I recommend checking for the presence of that key before trying to access it.

Reply:
Hmm this is not good, cause `was_informed_by` should have a value for all Workflow docs. I wonder which docs are causing that error.

Reply:
@eecavanna When I test this on a local yaml file of the database with the appropriate collections, the migrator works. I also checked Mongodb for any docs that don't have a `was_informed_by` slot in the `WorkflowExecution` subclasses, and I don't see that that case exists.

Reply:
I am wondering if this is an id-ing mismatch. I will ask @aclum

Reply:
Hi @brynnz22, I forgot that the database didn't actually match production when I ran this migrator. It did start out as a copy of the production database, but, by the time I ran this migrator, I had already run six other migrators:

1. `migrator_from_X_to_PR23`
2. `migrator_from_X_to_PR4`
3. `migrator_from_X_to_PR53`
4. `migrator_from_X_to_PR21`
5. `migrator_from_X_to_PR129`
6. `migrator_from_X_to_PR9`

So, the part of my original comment where I said "When I ran this migrator on a copy of the production database" was not accurate. I ran it on a copy of the production database that had undergone a six migrations.

---

My local database (the one I was running this migrator on) is still up and running, so I can get the problematic document. I will get it now and post it here in a few minutes for our future reference.

Reply:
Resolving this. This was due to this migrator needing to come before the `WorkflowChain` migration since it removes the `was_informed_by` slot. The meta issue has been updated with the correct ordering.